File: medium/1405_longest_happy_string/solution.py
import heapq


class Solution:
    def longestDiverseString(self, a: int, b: int, c: int) -> str:
        heap = []
        if a:
            heap.append((-a, "a"))
        if b:
            heap.append((-b, "b"))
        if c:
            heap.append((-c, "c"))

        heapq.heapify(heap)

        currCount = 0
        prevLetter = " "

        happyString = []

        while heap:
            count, letter = heapq.heappop(heap)

            if currCount == 2 and letter == prevLetter:
                if not heap:
                    break

                t = (count, letter)
                count, letter = heapq.heappop(heap)
                heapq.heappush(heap, t)

            happyString.append(letter)

            currCount = currCount + 1 if letter == prevLetter else 1
            prevLetter = letter

            if count + 1 < 0:
                heapq.heappush(heap, (count + 1, letter))

        return "".join(happyString)

File: medium/1405_longest_happy_string/solution.cpp
#include <queue>
#include <string>

class Solution {
public:
  std::string longestDiverseString(int a, int b, int c) {
    std::priority_queue<std::pair<int, char>> pq;
    if (a)
      pq.push({a, 'a'});
    if (b)
      pq.push({b, 'b'});
    if (c)
      pq.push({c, 'c'});

    int currCount = 0;
    char prevLetter = ' ';

    std::string happyString;

    while (!pq.empty()) {
      auto p = pq.top();
      pq.pop();

      int count = p.first;
      char letter = p.second;

      if (currCount == 2 && letter == prevLetter) {
        if (pq.empty())
          break;

        count = pq.top().first;
        letter = pq.top().second;

        pq.pop();
        pq.push(p);
      }

      happyString.push_back(letter);

      currCount = (letter == prevLetter) ? currCount + 1 : 1;
      prevLetter = letter;

      if (count - 1 > 0)
        pq.push({count - 1, letter});
    }

    return happyString;
  }
};
